ca/ca_test.go

```diff
@@ -207,261 +194,527 @@ func setup(t *testing.T) *testCtx {
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
+	}
 
-func issueCertificateSubTestSetup(t *testing.T) (*certificateAuthorityImpl, *mockSA) {
-	testCtx := setup(t)
-	sa := &mockSA{}
-	ca, err := NewCertificateAuthorityImpl(
-		sa,
-		mockSCTService{},
-		testCtx.pa,
-		testCtx.boulderIssuers,
-		testCtx.certProfiles,
-		testCtx.serialPrefix,
-		testCtx.maxNames,
-		testCtx.keyPolicy,
-		testCtx.logger,
-		testCtx.metrics,
-		testCtx.fc)
-	test.AssertNotError(t, err, "Failed to create CA")
+	for _, tc := range []struct {
+		name    string
+		csrPath string
+	}{
+		{
+			name:    "no names",
+			csrPath: "./testdata/no_names.der.csr",
+		},
+		{
+			name:    "too many names",
+			csrPath: "./testdata/too_many_names.der.csr",
+		},
+		{
+			name:    "short key",
+			csrPath: "./testdata/short_key.der.csr",
+		},
+		{
+			name:    "bad key algorithm",
+			csrPath: "./testdata/bad_algorithm.der.csr",
+		},
+		{
+			name:    "invalid signature",
+			csrPath: "./testdata/invalid_signature.der.csr",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
 
-	return ca, sa
+			_, err := ca.IssueCertificate(t.Context(), &capb.IssueCertificateRequest{
+				RegistrationID: 1, OrderID: 1,
+				Csr: mustRead(tc.csrPath), CertProfileName: "legacy",
+			})
+			if err == nil {
+				t.Fatalf("IssueCertificate(%q) succeeded, but want error", tc.csrPath)
+			}
+			if !errors.Is(err, berrors.BadCSR) {
+				t.Fatalf("IssueCertificate(%q) = %T, but want %T", tc.csrPath, err, berrors.BadCSR)
+			}
+		})
+	}
 }
 
-func issueCertificateSubTestIssuePrecertificate(t *testing.T, i *TestCertificateIssuance) {
-	cert := i.cert
+func TestIssueCertificate_ValidPastIssuer(t *testing.T) {
+	t.Parallel()
+	cargs := newCAArgs(t)
 
-	test.AssertEquals(t, cert.Subject.CommonName, "not-example.com")
+	// Limit ourselves to only having one ECDSA issuer, just in case they have
+	// different notAfter dates.
+	cargs.boulderIssuers = cargs.boulderIssuers[:3]
 
-	if len(cert.DNSNames) == 1 {
-		if cert.DNSNames[0] != "not-example.com" {
-			t.Errorf("Improper list of domain names %v", cert.DNSNames)
-		}
-		t.Errorf("Improper list of domain names %v", cert.DNSNames)
+	// Jump to a time just moments before the test issuer expire.
+	future := cargs.boulderIssuers[2].Cert.Certificate.NotAfter.Add(-1 * time.Hour)
+	cargs.clk.Set(future)
+
+	ca, err := cargs.make()
+	if err != nil {
+		t.Fatalf("making test ca: %s", err)
 	}
 
-	if len(cert.Subject.Country) > 0 {
-		t.Errorf("Subject contained unauthorized values: %v", cert.Subject)
+	_, err = ca.IssueCertificate(t.Context(), &capb.IssueCertificateRequest{
+		RegistrationID: 1, OrderID: 1,
+		Csr: ECDSACSR, CertProfileName: "legacy",
+	})
+	if err == nil {
+		t.Fatalf("IssueCertificate(notAfter > issuer.notAfter) succeeded, but want error")
+	}
+	if !errors.Is(err, berrors.InternalServer) {
+		t.Fatalf("IssueCertificate(notAfter > issuer.notAfter) = %T, but want %T", err, berrors.InternalServer)
```

Comment:
Rather than printing the type of the incorrect error let's print the whole error. I think that will result in quicker debugging.